File: src/main/kotlin/dev/jefy/connectpro/recommandation/application/RecommandationServiceImpl.java
package dev.jefy.connectpro.recommandation.application;

import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import dev.jefy.connectpro.management.domain.Category;
import dev.jefy.connectpro.management.domain.repositoty.CategoryRepository;
import dev.jefy.connectpro.management.domain.vo.CategoryId;
import dev.jefy.connectpro.portfolio.PortfolioClient;
import dev.jefy.connectpro.recommandation.domain.EventTracking;
import dev.jefy.connectpro.recommandation.domain.repository.EventTrackingRepository;
import dev.jefy.connectpro.recommandation.domain.vo.TargetType;
import dev.jefy.connectpro.user.UserClient;
import dev.jefy.connectpro.user.application.dtos.UserData;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */
@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class RecommandationServiceImpl implements RecommandationService {
    private final EventTrackingRepository eventTrackingRepo;
    private final UserClient userClient;
    private final PortfolioClient portfolioClient;
    private final CategoryRepository categoryRepository;

    @Override
    public List<CategoryId> getRecommendedCategories() {
        UserData currentUser = userClient.getCurrentUser();
        List<EventTracking> events = eventTrackingRepo.findAllByUserId(currentUser.id());
        
        if (events.isEmpty()) {
            return getPopularCategories();
        }

        Map<CategoryId, Integer> categoryScores = new HashMap<>();
        
        for (EventTracking event : events) {
            Optional<CategoryId> categoryIdOpt = getCategoryIdFromTarget(event);
            if (categoryIdOpt.isPresent()) {
                CategoryId categoryId = categoryIdOpt.get();
                int weight = getEventWeight(event);
                categoryScores.put(categoryId, categoryScores.getOrDefault(categoryId, 0) + weight);
            }
        }
        
        if (categoryScores.isEmpty()) {
            return getPopularCategories();
        }

        return categoryScores.entrySet().stream()
                .sorted((e1, e2) -> e2.getValue().compareTo(e1.getValue()))
                .map(Map.Entry::getKey)
                .limit(5)
                .collect(Collectors.toList());
    }

    private List<CategoryId> getPopularCategories() {
        return categoryRepository.findAll(PageRequest.of(0, 5))
                .getContent().stream()
                .map(Category::getId)
                .collect(Collectors.toList());
    }
    
    private Optional<CategoryId> getCategoryIdFromTarget(EventTracking event) {
        if (event.getTargetType() == TargetType.SERVICE || event.getTargetType() == TargetType.CONTACT_FOR_SERVICE) {
            return portfolioClient.getServiceCategoryId(event.getTargetId());
        } else if (event.getTargetType() == TargetType.JOB_POST || event.getTargetType() == TargetType.JOB_APPLICATION) {
            return portfolioClient.getJobPostCategoryId(event.getTargetId());
        } else if (event.getTargetType() == TargetType.LIKE) {
             // In current model, Like target might be a ServiceId (to be checked)
             return portfolioClient.getServiceCategoryId(event.getTargetId());
        } else if (event.getTargetType() == TargetType.REVIEW) {
             return portfolioClient.getServiceCategoryId(event.getTargetId());
        }
        return Optional.empty();
    }
    
    private int getEventWeight(EventTracking event) {
        return switch (event.getEventType()) {
            case LIKE -> 5;
            case REVIEW -> 4;
            case CLICK -> 2;
            case VIEW -> 1;
        };
    }
}

File: src/main/kotlin/dev/jefy/connectpro/recommandation/application/RecommandationServiceImpl.kt
package dev.jefy.connectpro.recommandation.application

import dev.jefy.connectpro.management.domain.Category
import dev.jefy.connectpro.management.domain.repository.CategoryRepository
import dev.jefy.connectpro.management.domain.vo.CategoryId
import dev.jefy.connectpro.portfolio.PortfolioClient
import dev.jefy.connectpro.recommandation.domain.EventTracking
import dev.jefy.connectpro.recommandation.domain.repository.EventTrackingRepository
import dev.jefy.connectpro.recommandation.domain.vo.EventType
import dev.jefy.connectpro.recommandation.domain.vo.TargetType
import dev.jefy.connectpro.user.UserClient
import dev.jefy.connectpro.user.application.dtos.UserData
import dev.jefy.connectpro.user.domain.vo.UserId
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class RecommandationServiceImpl(
    private val eventTrackingRepo: EventTrackingRepository,
    private val userClient: UserClient,
    private val portfolioClient: PortfolioClient,
    private val categoryRepository: CategoryRepository
) : RecommandationService {

    override fun getRecommendedCategories(): List<CategoryId> {
        val currentUser: UserData = userClient.getCurrentUser()
        val events = eventTrackingRepo.findAllByUserId(UserId.of(currentUser.id))

        if (events.isEmpty()) {
            return getPopularCategories()
        }

        val categoryScores = mutableMapOf<CategoryId, Int>()

        events.forEach { event ->
            val categoryIdOpt = getCategoryIdFromTarget(event)
            categoryIdOpt?.let { categoryId ->
                val weight = getEventWeight(event)
                categoryScores[categoryId] = categoryScores.getOrDefault(categoryId, 0) + weight
            }
        }

        if (categoryScores.isEmpty()) {
            return getPopularCategories()
        }

        return categoryScores.entries
            .sortedByDescending { it.value }
            .map { it.key }
            .take(5)
    }

    private fun getPopularCategories(): List<CategoryId> {
        return categoryRepository.findAll(PageRequest.of(0, 5))
            .content
            .map(Category::id)
    }

    private fun getCategoryIdFromTarget(event: EventTracking): CategoryId? {
        return when (event.targetType) {
            TargetType.SERVICE, TargetType.CONTACT_FOR_SERVICE -> portfolioClient.getServiceCategoryId(event.targetId).orElse(null)
            TargetType.JOB_POST, TargetType.JOB_APPLICATION -> portfolioClient.getJobPostCategoryId(event.targetId).orElse(null)
            TargetType.LIKE -> portfolioClient.getServiceCategoryId(event.targetId).orElse(null)
            TargetType.REVIEW -> portfolioClient.getServiceCategoryId(event.targetId).orElse(null)
            else -> null
        }
    }

    private fun getEventWeight(event: EventTracking): Int {
        return when (event.eventType) {
            EventType.LIKE -> 5
            EventType.REVIEW -> 4
            EventType.CLICK -> 2
            EventType.VIEW -> 1
        }
    }
}